### ISSUE TITLE
citygml2cityjson: appearances are not converted

### ISSUE THREAD

**Issue by hugoledoux** [citygml4j/citygml4j#11]
that is, the jpg files are not, but the json code for them is... or it's not clear to me what is happening. If I setup the `SimpleTextureFileHandler` then no `appearances` in json is created, if I skip it then yes. But the textures are never copied anywhere. (my java skills are pretty bad, I guess I could debug, but one thing at a time, I already managed to run all things correctly and write my own code based on citygml4j today!).

That file: http://3dre.gw.rotterdam.nl/3D/3-20-Delfshaven.zip
would it be because the paths for textures have backslashes instead of slashes?
`<app:imageURI>appearances\0320_0_13.jpg</app:imageURI>`

**Comment by clausnagel** [comment]
As you know, texture files can be freely organized for a CityGML dataset. For example, they can be stored in a single folder or in nested or distributed folders in which case they may even share the same filename. Moreover, `<app:imageURI>` may be a pointer into the local file system (relative or global) or may likewise reference a remote resource such as a web service. For CityJSON, all texture files have to go into a local folder named "appearances" relative to the CityJSON file. Moreover, CityJSON expects JPGs and PNGs whereas there is no such limitation in CityGML.
 
citygml4j does not solve this mapping problem as it may quickly become complex. Instead, it delegates the mapping to the developer by providing two simple `TextureFileHandler` interfaces - one for mapping from `<app:imageURI>` to the `"image"` property of a CityJSON texture object and one for the other way round. The first one defines a simple method:

```java
public String getImageFileName(String imageURI);
```

You can implement this interface and register the implementing class with a `CityJSONInputFactory`. Then the `getImageFileName` method of your class will be invoked for every `<app:imageURI>` found in the CityGML data and has to return a unique `"image"` property value. And it may/should also copy the texture file from the location referenced by `<app:imageURI>` to the "appearances" folder. If the referenced texture is neither a JPG nor a PNG, it could even apply an image type conversion.

If you do not register your own texture file handler, then a default one will be used. This default one, however, just strips the filename from the `<app:imageURI>` property and returns it. There is no check for duplicate filenames, for local or remote resources, no copying and no conversion. 

The `SimpleTextureFileHandler` provided in the samples might be a good starting point for your own texture file handler, but it is really very simple and tailored to the sample data. I could think of enhancing the default handler in an upcoming release such that it covers the most typical situations.

For your dataset, the default handler is already sufficient since the texture files are stored in an "appearances" folder next to the CityGML dataset. So all you have to do is convert the CityGML to CityJSON and put the resulting file into the same folder as the CityGML file. No need to register your own texture file handler with the `CityJSONInputFactory`. 

Hope this helps. Sorry for the lengthy post.

**Comment by hugoledoux** [comment]
I had registered the textures but they were not converted for this dataset, while for two others it just worked fine (same code).

So my question was why is that you think?

Your explanations make it clearer now though, thanks.

**Comment by clausnagel** [comment]
Well, are all your datasets in version 0.5 or do you perhaps use a previous version? 

And yes, also slash vs. backslash could make a difference... Are you running on Linux/MacOS?

**Comment by hugoledoux** [comment]
it was CityGML to CityJSON, so version is irrelevant, no?

I'm on macOS.

But basically: can you convert that dataset with your tool and keep/copy the textures to CityJSON? "/" or "\" shouldn't influence, should it?

**Comment by clausnagel** [comment]
Yes, I can convert the file without problems - both with the `SimpleTextureFileHandler` and the default texture file handler. The `"appearance"` object is created and contains `"textures"`. 

Here is the result (without the texture files): [3-20-DELFSHAVEN.zip](https://github.com/citygml4j/citygml4j/files/1559946/3-20-DELFSHAVEN.zip)

**Comment by hugoledoux** [comment]
hmmm then I’m doing smth wrong if Linux working.

**Comment by clausnagel** [comment]
Sorry Hugo, imprecise wording. I have not yet tested under Linux. 

Could you please try and change the following line in https://github.com/hugoledoux/city_gml_json_4j/blob/master/cg2cj/util/SimpleTextureFileHandler.java:
```java
line 62: Path source = baseDir.resolve(imageURI); 
```
into
```java
line 62: Path source = baseDir.resolve(imageURI.replace("\\", "/"));
```
I think your issue is related to the backslash. If this does not help though, then I will setup a Linux box to do more debugging.

**Comment by clausnagel** [comment]
Fixed in 9492f5f53bbeefcea8b0a3a1748f7fd975964b88
